File: Lab 4/q1.c

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to perform Merge Sort
long long merge(long long arr[], long long temp[], int left, int mid, int right) {
    int i = left, j = mid, k = left;
    long long comparisons = 0;

    while ((i <= mid - 1) && (j <= right)) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
        comparisons++;
    }

    while (i <= mid - 1) {
        temp[k++] = arr[i++];
    }
    while (j <= right) {
        temp[k++] = arr[j++];
    }

    for (i = left; i <= right; i++) {
        arr[i] = temp[i];
    }

    return comparisons;
}

long long mergeSort(long long arr[], long long temp[], int left, int right) {
    long long comparisons = 0;
    if (right > left) {
        int mid = (left + right) / 2;
        comparisons += mergeSort(arr, temp, left, mid);
        comparisons += mergeSort(arr, temp, mid + 1, right);
        comparisons += merge(arr, temp, left, mid + 1, right);
    }
    return comparisons;
}

int main() {
    FILE *inputFile, *outputFile;
    int choice, i, n;
    clock_t start, end;
    double cpu_time_used;
    
    printf("MAIN MENU (MERGE SORT)\n");
    printf("1. Ascending Data\n");
    printf("2. Descending Data\n");
    printf("3. Random Data\n");
    printf("4. Exit\n");
    
    printf("Enter option: ");
    scanf("%d", &choice);
    
    char *inputFilename, *outputFilename;
    
    switch (choice) {
        case 1:
            inputFilename = "inAsce.dat";
            outputFilename = "outMergeAsce.dat";
            break;
        case 2:
            inputFilename = "inDesc.dat";
            outputFilename = "outMergeDesc.dat";
            break;
        case 3:
            inputFilename = "inRand.dat";
            outputFilename = "outMergeRand.dat";
            break;
        case 4:
            exit(0);
            break;
        default:
            printf("Invalid option!\n");
            exit(1);
    }
    
    inputFile = fopen(inputFilename, "r");
    if (inputFile == NULL) {
        perror("Error opening input file");
        exit(1);
    }
    
    fscanf(inputFile, "%d", &n);
    long long arr[n], temp[n];
    for (i = 0; i < n; i++) {
        fscanf(inputFile, "%lld", &arr[i]);
    }
    fclose(inputFile);

    start = clock();
    long long comparisons = mergeSort(arr, temp, 0, n - 1);
    end = clock();

    outputFile = fopen(outputFilename, "w");
    if (outputFile == NULL) {
        perror("Error opening output file");
        exit(1);
    }
    
    fprintf(outputFile, "After Sorting:\n");
    for (i = 0; i < n; i++) {
        fprintf(outputFile, "%lld ", arr[i]);
    }
    fclose(outputFile);

    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC * 1e9; // Convert to nanoseconds
    
    printf("Number of Comparisons: %lld\n", comparisons);
    printf("Execution Time: %lf nanoseconds\n", cpu_time_used);

    return 0;
}

File: Lab 4/q1.java
import java.io.*;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class MergeSortMenu {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        boolean exit = false;
        
        while (!exit) {
            System.out.println("MAIN MENU (MERGE SORT)");
            System.out.println("1. Ascending Data");
            System.out.println("2. Descending Data");
            System.out.println("3. Random Data");
            System.out.println("4. Exit");
            System.out.print("Enter option: ");
            
            int option = scanner.nextInt();
            
            switch (option) {
                case 1:
                    performMergeSort("inAsce.dat", "outMergeAsce.dat");
                    break;
                case 2:
                    performMergeSort("inDesc.dat", "outMergeDesc.dat");
                    break;
                case 3:
                    performMergeSort("inRand.dat", "outMergeRand.dat");
                    break;
                case 4:
                    exit = true;
                    break;
                default:
                    System.out.println("Invalid option. Please choose a valid option.");
            }
        }
        
        System.out.println("Exiting...");
    }
    
    public static void performMergeSort(String inputFile, String outputFile) {
        try {
            ArrayList<Integer> data = readFromFile(inputFile);
            
            long startTime = System.nanoTime();
            long comparisons = mergeSort(data, 0, data.size() - 1);
            long endTime = System.nanoTime();
            
            writeToFile(outputFile, data);
            
            System.out.println("Before Sorting: " + data);
            System.out.println("After Sorting: " + data);
            System.out.println("Number of Comparisons: " + comparisons);
            System.out.println("Execution Time: " + (endTime - startTime) + " nanoseconds");
        } catch (IOException e) {
            System.out.println("Error reading/writing file: " + e.getMessage());
        }
    }
    
    public static ArrayList<Integer> readFromFile(String fileName) throws IOException {
        ArrayList<Integer> data = new ArrayList<>();
        BufferedReader reader = new BufferedReader(new FileReader(fileName));
        String line;
        
        while ((line = reader.readLine()) != null) {
            String[] tokens = line.split(" ");
            for (String token : tokens) {
                data.add(Integer.parseInt(token));
            }
        }
        
        reader.close();
        return data;
    }
    
    public static void writeToFile(String fileName, ArrayList<Integer> data) throws IOException {
        BufferedWriter writer = new BufferedWriter(new FileWriter(fileName));
        for (int num : data) {
            writer.write(num + " ");
        }
        writer.close();
    }
    
    public static long mergeSort(ArrayList<Integer> arr, int left, int right) {
        long comparisons = 0;
        
        if (left < right) {
            int mid = (left + right) / 2;
            
            comparisons += mergeSort(arr, left, mid);
            comparisons += mergeSort(arr, mid + 1, right);
            
            comparisons += merge(arr, left, mid, right);
        }
        
        return comparisons;
    }
    
    public static long merge(ArrayList<Integer> arr, int left, int mid, int right) {
        int[] temp = new int[right - left + 1];
        int i = left, j = mid + 1, k = 0;
        long comparisons = 0;
        
        while (i <= mid && j <= right) {
            if (arr.get(i) <= arr.get(j)) {
                temp[k++] = arr.get(i++);
            } else {
                temp[k++] = arr.get(j++);
            }
            comparisons++;
        }
        
        while (i <= mid) {
            temp[k++] = arr.get(i++);
        }
        
        while (j <= right) {
            temp[k++] = arr.get(j++);
        }
        
        for (i = left, k = 0; i <= right; i++, k++) {
            arr.set(i, temp[k]);
        }
        
        return comparisons;
    }
}
